File: apps/api/src/schemas/intake.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from src.enums import SourcePlatformEnum
from src.schemas.candidates import FilterConfigRequest


class IntakeDiscoverRequest(BaseModel):
    workspace_id: UUID | None = None
    profile_url: str
    source_platform: SourcePlatformEnum = SourcePlatformEnum.DOUYIN
    preset_name: str | None = None
    filter_config: FilterConfigRequest | None = None
    persist: bool = True
    force_live_refresh: bool = False
    douyin_account_connection_id: UUID | None = None

    @model_validator(mode="after")
    def validate_thresholds(self) -> "IntakeDiscoverRequest":
        if self.filter_config is None:
            return self
        config = self.filter_config
        if config.min_views is not None and config.max_views is not None and config.min_views > config.max_views:
            raise ValueError("min_views must be less than or equal to max_views")
        if config.min_likes is not None and config.max_likes is not None and config.min_likes > config.max_likes:
            raise ValueError("min_likes must be less than or equal to max_likes")
        if config.min_comments is not None and config.max_comments is not None and config.min_comments > config.max_comments:
            raise ValueError("min_comments must be less than or equal to max_comments")
        if config.min_shares is not None and config.max_shares is not None and config.min_shares > config.max_shares:
            raise ValueError("min_shares must be less than or equal to max_shares")
        if (
            config.min_duration_seconds is not None
            and config.max_duration_seconds is not None
            and config.min_duration_seconds > config.max_duration_seconds
        ):
            raise ValueError("min_duration_seconds must be less than or equal to max_duration_seconds")
        if (
            config.min_engagement_rate is not None
            and config.max_engagement_rate is not None
            and config.min_engagement_rate > config.max_engagement_rate
        ):
            raise ValueError("min_engagement_rate must be less than or equal to max_engagement_rate")
        if config.start_date is not None and config.end_date is not None and config.start_date > config.end_date:
            raise ValueError("start_date must be before or equal to end_date")
        return self


class IntakeReadyCheckRequest(BaseModel):
    workspace_id: UUID | None = None
    profile_url: str | None = None
    douyin_account_connection_id: UUID | None = None


class IntakeReadyCheckResponse(BaseModel):
    diagnostics_id: str
    readiness_status: str
    safe_to_run_intake_now: bool
    selected_account_id: UUID | None = None
    selected_account_label: str | None = None
    resolved_account_id: UUID | None = None
    resolved_account_label: str | None = None
    account_selection_mode: str | None = None
    account_selection_reason: str | None = None
    account_fallback_notice: str | None = None
    account_health: str | None = None
    browser_profile_status: str | None = None
    browser_profile_available: bool = False
    browser_reopen_needed: bool = False
    browser_reopen_attempted: bool = False
    browser_reopen_result: str | None = None
    intended_fetch_path: str | None = None
    fallback_allowed: bool = False
    recommended_action: str
    recommended_action_label: str
    summary_message: str
    preflight_cached: bool = False
    watchdog_result: str | None = None
    watchdog_status: str | None = None
    watchdog_reason: str | None = None
    preflight_result: str | None = None
    fetch_readiness_category: str | None = None
    preflight_failure_code: str | None = None
    preflight_failure_message: str | None = None
    challenge_state: str | None = None
    challenge_category: str | None = None
    challenge_count: int | None = None
    challenge_cooldown_until: datetime | None = None
    challenge_recommended_next_action: str | None = None
    profile_quarantine_state: str = "active_preferred"
    profile_quarantine_reason: str | None = None
    profile_quarantine_detected: bool = False
    profile_quarantine_recommended_next_action: str | None = None
    profile_quarantine_blocks_primary_flow: bool = False
    profile_quarantine_replaced_by_account_id: UUID | None = None
    profile_quarantine_clean_profile_recommendation: str | None = None
    profile_url: str | None = None


class IntakeDiscoverResponse(BaseModel):
    success: bool = True
    diagnostics_id: str
    source_profile_id: UUID
    crawl_session_id: UUID | None
    submitted_profile_url: str
    normalized_profile_identifier: str | None
    videos_discovered_count: int = 0
    videos_created_count: int = 0
    videos_updated_count: int = 0
    candidates_total_count: int = 0
    candidates_matched_count: int = 0
    candidates_rejected_count: int = 0
    candidate_results_count: int = 0
    filters_applied_summary: dict = Field(default_factory=dict)
    unsupported_filters_ignored: list[str] = Field(default_factory=list)
    fetch_mode: str = "live_fetch"
    used_existing_profile: bool = False
    douyin_account_connection_id: UUID | None = None
    selected_douyin_account_connection_id: UUID | None = None
    resolved_douyin_account_connection_id: UUID | None = None
    douyin_account_selection_mode: str | None = None
    douyin_account_selection_reason: str | None = None
    douyin_account_fallback_notice: str | None = None
    fetch_stage: str | None = None
    fetch_stage_code: str | None = None
    fetch_stage_message: str | None = None
    parser_strategy: str | None = None
    fetch_execution_path: str | None = None
    fallback_from_execution_path: str | None = None
    strategy_policy: str | None = None
    primary_execution_path: str | None = None
    http_fallback_attempted: bool | None = None
    http_fallback_reason: str | None = None
    preflight_ran: bool = False
    preflight_result: str | None = None
    fetch_readiness_category: str | None = None
    selected_fetch_path: str | None = None
    browser_reopen_attempted: bool | None = None
    browser_reopen_result: str | None = None
    preflight_failure_code: str | None = None
    preflight_cached: bool | None = None
    watchdog_result: str | None = None
    watchdog_status: str | None = None
    watchdog_reason: str | None = None
    runtime_reconciled: bool | None = None
    videos_normalized_count: int = 0
    videos_persisted_count: int = 0
    next_suggested_route: str = "/review-board?fresh=1"
    warning: str | None = Field(default=None)
    discovered_at: datetime


class IntakeSavedPresetPayload(BaseModel):
    profile_url: str
    preset_name: str | None = None
    filter_config: FilterConfigRequest | None = None
    force_live_refresh: bool = False
    douyin_account_connection_id: UUID | None = None


class IntakeSavedPresetCreateRequest(IntakeSavedPresetPayload):
    workspace_id: UUID | None = None
    name: str = Field(min_length=1, max_length=120)
    notes: str | None = None


class IntakeSavedPresetUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    profile_url: str | None = None
    preset_name: str | None = None
    filter_config: FilterConfigRequest | None = None
    force_live_refresh: bool | None = None
    douyin_account_connection_id: UUID | None = None
    notes: str | None = None


class IntakeSavedPresetResponse(BaseModel):
    id: UUID
    workspace_id: UUID
    name: str
    profile_url: str
    preset_name: str | None
    filter_config: dict = Field(default_factory=dict)
    force_live_refresh: bool
    douyin_account_connection_id: UUID | None
    notes: str | None
    created_at: datetime
    updated_at: datetime


class IntakeSavedPresetListResponse(BaseModel):
    presets: list[IntakeSavedPresetResponse]


class IntakeRecentProfileResponse(BaseModel):
    source_profile_id: UUID
    profile_url: str
    normalized_profile_identifier: str | None = None
    display_name: str | None = None
    last_crawled_at: datetime | None = None


class IntakeLatestSuccessShortcutResponse(BaseModel):
    crawl_session_id: UUID
    source_profile_id: UUID | None = None
    submitted_profile_url: str | None = None
    normalized_profile_identifier: str | None = None
    finished_at: datetime | None = None
    videos_discovered_count: int = 0


class IntakeBootstrapResponse(BaseModel):
    workspace_id: UUID
    saved_presets: list[IntakeSavedPresetResponse]
    recent_profiles: list[IntakeRecentProfileResponse]
    latest_success_shortcuts: list[IntakeLatestSuccessShortcutResponse]


class IntakeRunSummaryResponse(BaseModel):
    crawl_session_id: UUID
    source_profile_id: UUID | None = None
    submitted_profile_url: str | None = None
    normalized_profile_identifier: str | None = None
    source_profile_display_name: str | None = None
    status: str
    fetch_mode: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    duration_seconds: int | None = None
    videos_discovered_count: int = 0
    videos_created_count: int = 0
    videos_updated_count: int = 0
    candidates_total_count: int = 0
    candidates_matched_count: int = 0
    error_code: str | None = None
    error_message: str | None = None
    fetch_observability: dict = Field(default_factory=dict)


class IntakeRunListResponse(BaseModel):
    runs: list[IntakeRunSummaryResponse]


class IntakeTroubleshootingSummaryResponse(BaseModel):
    category: str
    severity: str
    why: str
    recommended_actions: list[str] = Field(default_factory=list)


class IntakeRunDetailResponse(IntakeRunSummaryResponse):
    troubleshooting: IntakeTroubleshootingSummaryResponse


class IntakeRunCompareResponse(BaseModel):
    left: IntakeRunSummaryResponse
    right: IntakeRunSummaryResponse
    status_changed: bool
    duration_seconds_delta: int | None = None
    videos_discovered_delta: int = 0
    videos_created_delta: int = 0
    videos_updated_delta: int = 0
    error_code_changed: bool
    left_error_code: str | None = None
    right_error_code: str | None = None
    left_candidates_total: int = 0
    right_candidates_total: int = 0
    candidates_total_delta: int = 0
    left_candidates_matched: int = 0
    right_candidates_matched: int = 0
    candidates_matched_delta: int = 0

File: apps/web/src/types/intake.ts
export type IntakeFilterConfig = {
  date_mode?: "absolute_range";
  start_date?: string;
  end_date?: string;
  min_views?: number;
  max_views?: number;
  min_likes?: number;
  max_likes?: number;
  min_comments?: number;
  max_comments?: number;
  min_shares?: number;
  max_shares?: number;
  min_duration_seconds?: number;
  max_duration_seconds?: number;
  min_engagement_rate?: number;
  max_engagement_rate?: number;
  has_speech?: boolean;
  max_text_density?: "low" | "medium" | "high";
  exclude_heavy_watermark?: boolean;
  exclude_high_processing_complexity?: boolean;
  exclude_high_copyright_risk?: boolean;
  sort: "score_desc";
  limit: number;
  offset: number;
};

export type IntakeDiscoverRequest = {
  profile_url: string;
  preset_name: string | null;
  filter_config: IntakeFilterConfig;
  persist: boolean;
  force_live_refresh: boolean;
  douyin_account_connection_id: string | null;
};

export type IntakeReadyCheckRequest = {
  workspace_id?: string | null;
  profile_url?: string | null;
  douyin_account_connection_id?: string | null;
};

export type IntakeReadyCheckResponse = {
  diagnostics_id: string;
  readiness_status: "READY" | "READY_AFTER_REOPEN" | "FALLBACK_READY" | "CHALLENGE_BLOCKED" | "PROFILE_QUARANTINED" | "NOT_READY" | string;
  safe_to_run_intake_now: boolean;
  selected_account_id: string | null;
  selected_account_label: string | null;
  resolved_account_id: string | null;
  resolved_account_label: string | null;
  account_selection_mode: string | null;
  account_selection_reason: string | null;
  account_fallback_notice: string | null;
  account_health: string | null;
  browser_profile_status: string | null;
  browser_profile_available: boolean;
  browser_reopen_needed: boolean;
  browser_reopen_attempted: boolean;
  browser_reopen_result: string | null;
  intended_fetch_path: string | null;
  fallback_allowed: boolean;
  recommended_action: string;
  recommended_action_label: string;
  summary_message: string;
  preflight_cached: boolean;
  watchdog_result: string | null;
  watchdog_status: string | null;
  watchdog_reason: string | null;
  preflight_result: string | null;
  fetch_readiness_category: string | null;
  preflight_failure_code: string | null;
  preflight_failure_message: string | null;
  challenge_state: string | null;
  challenge_category: string | null;
  challenge_count: number | null;
  challenge_cooldown_until: string | null;
  challenge_recommended_next_action: string | null;
  profile_quarantine_state: string;
  profile_quarantine_reason: string | null;
  profile_quarantine_detected: boolean;
  profile_quarantine_recommended_next_action: string | null;
  profile_quarantine_blocks_primary_flow: boolean;
  profile_quarantine_replaced_by_account_id: string | null;
  profile_quarantine_clean_profile_recommendation: string | null;
  profile_url: string | null;
};

export type IntakeDiscoverResponse = {
  success: boolean;
  diagnostics_id: string;
  source_profile_id: string;
  crawl_session_id: string | null;
  submitted_profile_url: string;
  normalized_profile_identifier: string | null;
  videos_discovered_count: number;
  videos_created_count: number;
  videos_updated_count: number;
  candidates_total_count: number;
  candidates_matched_count: number;
  candidates_rejected_count: number;
  candidate_results_count: number;
  filters_applied_summary: Record<string, unknown>;
  unsupported_filters_ignored: string[];
  fetch_mode: string;
  used_existing_profile: boolean;
  douyin_account_connection_id: string | null;
  selected_douyin_account_connection_id: string | null;
  resolved_douyin_account_connection_id: string | null;
  douyin_account_selection_mode: string | null;
  douyin_account_selection_reason: string | null;
  douyin_account_fallback_notice: string | null;
  fetch_stage: string | null;
  fetch_stage_code: string | null;
  fetch_stage_message: string | null;
  parser_strategy: string | null;
  fetch_execution_path: string | null;
  fallback_from_execution_path: string | null;
  strategy_policy: string | null;
  primary_execution_path: string | null;
  http_fallback_attempted: boolean | null;
  http_fallback_reason: string | null;
  preflight_ran: boolean;
  preflight_result: string | null;
  fetch_readiness_category: string | null;
  selected_fetch_path: string | null;
  browser_reopen_attempted: boolean | null;
  browser_reopen_result: string | null;
  preflight_failure_code: string | null;
  preflight_cached: boolean | null;
  watchdog_result: string | null;
  watchdog_status: string | null;
  watchdog_reason: string | null;
  runtime_reconciled: boolean | null;
  videos_normalized_count: number;
  videos_persisted_count: number;
  next_suggested_route: string;
  warning: string | null;
  discovered_at: string;
};

export type IntakeSavedPresetPayload = {
  profile_url: string;
  preset_name: string | null;
  filter_config: IntakeFilterConfig;
  force_live_refresh: boolean;
  douyin_account_connection_id: string | null;
};

export type IntakeSavedPresetCreateRequest = IntakeSavedPresetPayload & {
  workspace_id?: string | null;
  name: string;
  notes?: string | null;
};

export type IntakeSavedPresetUpdateRequest = {
  name?: string;
  profile_url?: string;
  preset_name?: string | null;
  filter_config?: IntakeFilterConfig;
  force_live_refresh?: boolean;
  douyin_account_connection_id?: string | null;
  notes?: string | null;
};

export type IntakeSavedPresetResponse = {
  id: string;
  workspace_id: string;
  name: string;
  profile_url: string;
  preset_name: string | null;
  filter_config: Record<string, unknown>;
  force_live_refresh: boolean;
  douyin_account_connection_id: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type IntakeSavedPresetListResponse = {
  presets: IntakeSavedPresetResponse[];
};

export type IntakeRecentProfileResponse = {
  source_profile_id: string;
  profile_url: string;
  normalized_profile_identifier: string | null;
  display_name: string | null;
  last_crawled_at: string | null;
};

export type IntakeLatestSuccessShortcutResponse = {
  crawl_session_id: string;
  source_profile_id: string | null;
  submitted_profile_url: string | null;
  normalized_profile_identifier: string | null;
  finished_at: string | null;
  videos_discovered_count: number;
};

export type IntakeBootstrapResponse = {
  workspace_id: string;
  saved_presets: IntakeSavedPresetResponse[];
  recent_profiles: IntakeRecentProfileResponse[];
  latest_success_shortcuts: IntakeLatestSuccessShortcutResponse[];
};

export type IntakeRunSummaryResponse = {
  crawl_session_id: string;
  source_profile_id: string | null;
  submitted_profile_url: string | null;
  normalized_profile_identifier: string | null;
  source_profile_display_name: string | null;
  status: string;
  fetch_mode: string | null;
  started_at: string | null;
  finished_at: string | null;
  duration_seconds: number | null;
  videos_discovered_count: number;
  videos_created_count: number;
  videos_updated_count: number;
  candidates_total_count: number;
  candidates_matched_count: number;
  error_code: string | null;
  error_message: string | null;
};

export type IntakeRunListResponse = {
  runs: IntakeRunSummaryResponse[];
};

export type IntakeTroubleshootingSummaryResponse = {
  category: string;
  severity: string;
  why: string;
  recommended_actions: string[];
};

export type IntakeRunDetailResponse = IntakeRunSummaryResponse & {
  troubleshooting: IntakeTroubleshootingSummaryResponse;
};

export type IntakeRunCompareResponse = {
  left: IntakeRunSummaryResponse;
  right: IntakeRunSummaryResponse;
  status_changed: boolean;
  duration_seconds_delta: number | null;
  videos_discovered_delta: number;
  videos_created_delta: number;
  videos_updated_delta: number;
  error_code_changed: boolean;
  left_error_code: string | null;
  right_error_code: string | null;
  left_candidates_total: number;
  right_candidates_total: number;
  candidates_total_delta: number;
  left_candidates_matched: number;
  right_candidates_matched: number;
  candidates_matched_delta: number;
};

export type IntakeFormValues = {
  profileUrl: string;
  dateFrom: string;
  dateTo: string;
  minViews: string;
  maxViews: string;
  minLikes: string;
  maxLikes: string;
  minComments: string;
  maxComments: string;
  minShares: string;
  maxShares: string;
  minDurationSeconds: string;
  maxDurationSeconds: string;
  minEngagementRate: string;
  maxEngagementRate: string;
  hasSpeech: "any" | "yes" | "no";
  maxTextDensity: "" | "low" | "medium" | "high";
  excludeHeavyWatermark: boolean;
  excludeHighProcessingComplexity: boolean;
  excludeHighCopyrightRisk: boolean;
  forceLiveRefresh: boolean;
  douyinAccountConnectionId: string;
  presetName: string;
};

export type IntakeValidationErrors = Partial<Record<keyof IntakeFormValues | "form", string>>;

export type RecentIntakeSetup = {
  profileUrl: string;
  presetName: string;
  discoveredAt: string;
};
